src/riak_kv_index_hashtree.erl

```diff
@@ -507,31 +507,53 @@ determine_data_root() ->
 %%      then we immediately use version in capabilities. Otherwise check if capabilities
 %%      has flipped to a version yet and if it has, check if we've already upgraded by
 %%      looking for versioned AAE directory.
--spec determine_version(undefined | list(), non_neg_integer(), list()) -> legacy | non_neg_integer().
+-spec determine_version(list(), index(), list()) -> version().
 determine_version(Root, Index, Opts) ->
+    case force_upgrade(Opts) of
+        true ->
+            get_cap_hash_version();
+        _ ->
+            find_version(Root, Index)
+    end.
+
+-spec force_upgrade(list()) -> boolean().
+force_upgrade(Opts) ->
+    check_upgrade_opts(Opts, check_upgrade_env()).
+
+-spec check_upgrade_opts(list(), boolean()) -> boolean().
+check_upgrade_opts(_Opts, true) ->
+    true;
+check_upgrade_opts(Opts, _) ->
+    lists:member(upgrade, Opts).
+
+-spec check_upgrade_env() -> boolean().
+check_upgrade_env() ->
     case application:get_env(riak_kv, force_hashtree_upgrade) of
```

Comment:
Sorry - one more. How about using the 3-arity version of get_env and providing false as the default, rather than having to deal with the {ok, true} - can just match on true and _ at that point (in case someone put something stupid other than true in there).